src/main/java/org/spongepowered/api/command/CommandCause.java

```diff
@@ -0,0 +1,187 @@
+ * provide hints as to what the implementation will select.</p>
+ */
+public interface CommandCause {
+
+    /**
+     * Creates a {@link CommandCause} from the provided {@link Cause}
+     *
+     * @param cause The {@link Cause}
+     * @return The {@link CommandCause}
+     */
+    static CommandCause create(Cause cause) {
+        return Sponge.getRegistry().requireFactory(Factory.class).create(cause);
+    }
+
+    /**
+     * Gets the {@link Cause} of the command invocation.
+     *
+     * @return The cause of the invocation.
+     */
+    Cause getCause();
+
+    /**
+     * Get the {@link Subject} that will be selected for permission checks
+     * during command execution (by default).
+     *
+     * <p>The {@link Subject} will be selected in the following way from the
+     * {@link Cause} in {@link #getCause()}:</p>
+     *
+     * <ul>
+     *    <li>The {@link EventContextKeys#SUBJECT}, if any</li>
+     *    <li>The <strong>first</strong> {@link Subject} in the {@link Cause}
+     *    </li>
+     *    <li>The {@link SystemSubject} if no subject exists within the cause
+     *    </li>
+     * </ul>
+     *
+     * <p><strong>Note:</strong> while it might be tempting to use this as the
+     * invoker of the command, the {@link Cause#root()} and this might be
+     * different. Command executors should generally use the root of the
+     * {@link Cause} as the target of their command.</p>
+     *
+     * @return The {@link Subject} responsible, if any.
+     */
+    default Subject getSubject() {
+        return getCause().getContext()
+                .get(EventContextKeys.SUBJECT)
+                .orElseGet(() -> getCause().first(Subject.class).orElseGet(Sponge::getSystemSubject));
+    }
+
+    /**
+     * Gets the {@link MessageReceiver} that should be the target for any
+     * messages sent by the command (by default).
+     *
+     * <p>The {@link MessageReceiver} will be selected in the following way
+     * from the {@link Cause} in {@link #getCause()}:</p>
+     *
+     * <ul>
+     *    <li>The {@link EventContextKeys#MESSAGE_TARGET}, if any</li>
+     *    <li>A message channel containing the <strong>first</strong>
+     *    {@link MessageReceiver} in the {@link Cause}</li>
+     *    <li>The SystemSubject {@link MessageReceiver}</li>
+     * </ul>
+     *
+     * <p>Note that this returns a {@link MessageReceiver} and it may not what
+     * may be thought of as a traditional entity executing the command.
+     * For the object that invoked the command, check the
+     * {@link Cause#root()} of the {@link #getCause()}.</p>
+     *
+     * @return The {@link MessageReceiver} to send any messages to.
+     */
+    default MessageReceiver getMessageReceiver() {
+        return getCause().getContext()
+                .get(EventContextKeys.MESSAGE_TARGET)
+                .orElseGet(() -> getCause().first(MessageReceiver.class).orElseGet(Sponge::getSystemSubject));
```

Comment:
Wouldn't it be better to return a `Optional` here. Right now falling back to the "system subject" will result in logging all messages to the console, e.g. for command blocks, signs. Which isn't desired in most cases IMO. You could make your own method, but that's less convenient. Adding a separate method which falls back to the system subject is also an option.

Reply:
> Which isn't desired in most cases IMO. 

And won't normally happen in this system either. Command blocks are message receivers anyway so messages will go to them. Signs are explicity _not_ `MessageReceivers` and I would imagine that a sign would be triggered by a player, which would also be in the cause and so the message would end up with the player anyway. In the cases you've described the message generally _wouldn't_ go to the system subject - and I expect that to be the exception, rather than the norm.

This is intended to catch the scenario that a plugin executes a command that wasn't caused by a player, e.g. during a scheduled task, where a command would end up going to the "console" (as that is the command source you picked).

Finally, getting a `MessageReceiver` is a fundamental thing that command executors will do - I suspect wrapping it in an `Optional` is going to frustrate more than help, and that the vast majority of players will just add a `orElseGet(() -> Sponge.getSystemSubject())`. This is intended to point them at the appropriate receiver - if all else fails, that's the `SystemSubject`. If I add another optional method, you can bet no-one is going to use that in favour of this.